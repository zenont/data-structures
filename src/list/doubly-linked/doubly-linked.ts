export type MaybeDoublyLinkedListItem<T> = DoublyLinkedListItem<T> | undefined

/**
 * DoublyLinkedListItem.
 * Contains a prev and next reference.
 * @export
 * @class DoublyLinkedListItem
 * @template T
 */
export class DoublyLinkedListItem<T> {
	public prev: MaybeDoublyLinkedListItem<T> = undefined
	public next: MaybeDoublyLinkedListItem<T> = undefined
	/**
	 * Creates an instance of DoublyLinkedListItem.
	 * @param {T} value
	 * @param {MaybeDoublyLinkedListItem<T>} [prev]
	 * @param {MaybeDoublyLinkedListItem<T>} [next]
	 * @memberof DoublyLinkedListItem
	 */
	constructor(public value: T, prev?: MaybeDoublyLinkedListItem<T>, next?: MaybeDoublyLinkedListItem<T>) {
		this.prev = prev
		this.next = next
	}
}

/**
 * Doubly Linked list class.
 * Creates an instance of @class DoublyLinkedList
 * @export
 * @class DoublyLinkedList
 * @template T
 */
export class DoublyLinkedList<T> {
	/**
	 * Creates an instance of @class DoublyLinkedList from an array.
	 *
	 * @static
	 * @template T
	 * @param {...T[]} values
	 * @returns {DoublyLinkedList<T>}
	 * @memberof DoublyLinkedList
	 */
	public static of<T>(...values: T[]): DoublyLinkedList<T> {
		return new DoublyLinkedList<T>(...values)
	}

	private _head: MaybeDoublyLinkedListItem<T> = undefined
	private _tail: MaybeDoublyLinkedListItem<T> = undefined

	/**
	 * Creates an instance of DoublyLinkedList.
	 * @param {...T[]} values
	 * @memberof DoublyLinkedList
	 */
	constructor(...values: T[]) {
		values.forEach(value => this.push(value))
	}

	/**
	 * Push a value into the list.
	 * The value will create a new node at the tail of the list.
	 * @param {T} value
	 * @returns {DoublyLinkedListItem<T>}
	 * @memberof DoublyLinkedList
	 */
	public push(value: T): DoublyLinkedListItem<T> {
		const item = new DoublyLinkedListItem(value)
		if (this._head == null) {
			this._head = item
			this._tail = item
		} else {
			if (this._tail == null) throw new Error('Tail is undefined')
			item.prev = this._tail
			item.next = undefined
			this._tail.next = item
			this._tail = item
		}
		return item
	}

	/**
	 * Push a value after an item in the list.
	 * The value will create a new node after the passed item.
	 * @param {DoublyLinkedListItem<T>} item
	 * @param {T} value
	 * @returns {DoublyLinkedListItem<T>}
	 * @memberof DoublyLinkedList
	 */
	public pushAfter(item: DoublyLinkedListItem<T>, value: T): DoublyLinkedListItem<T> {
		if (item == null) throw new Error('The item you are trying to push after is null or undefined')

		const itemToPush = new DoublyLinkedListItem(value, item)
		if (item.next != null) {
			itemToPush.next = item.next
			item.next.prev = itemToPush
		}
		item.next = itemToPush
		if (item === this._tail) {
			this._tail = itemToPush
		}
		return itemToPush
	}

	/**
	 * Pops a value from the list.
	 * The value will be popped from the tail of the list.
	 * @returns {(T | undefined)}
	 * @memberof DoublyLinkedList
	 */
	public pop(): T | undefined {
		const tail = this.tail()
		if (tail == null) return undefined

		this.remove(tail)
		return tail.value
	}

	/**
	 * Removes an item from the list
	 *
	 * @param {DoublyLinkedListItem<T>} item
	 * @returns {DoublyLinkedList<T>}
	 * @memberof DoublyLinkedList
	 */
	public remove(item: DoublyLinkedListItem<T>): DoublyLinkedList<T> {
		if (item == null || this._head == null) return this

		if (item.prev != null) {
			item.prev.next = item.next
		} else {
			// the item to be removed is the head, so 'next' has to be set as the new head
			this._head = item.next
			if (this._head != null) {
				// if there is a head, then we make sure the prev is undefined
				this._head.prev = undefined
			}
		}

		if (item.next != null) {
			item.next.prev = item.prev
		} else {
			// the item to be removed is the tail, so 'prev' has to be set as the new tail
			this._tail = item.prev
			if (this._tail != null) {
				// if there is a new tail, we make sure the next is undefined
				this._tail.next = undefined
			}
		}
		return this
	}

	/**
	 * Gets the head of the list.
	 * Expect undefined if the list is empty.
	 * @returns {MaybeDoublyLinkedListItem<T>}
	 * @memberof DoublyLinkedList
	 */
	public head(): MaybeDoublyLinkedListItem<T> {
		return this._head
	}

	/**
	 * Gets the tail of the list.
	 * Expect undefined if the list is empty.
	 * @returns {MaybeDoublyLinkedListItem<T>}
	 * @memberof DoublyLinkedList
	 */
	public tail(): MaybeDoublyLinkedListItem<T> {
		return this._tail
	}

	/**
	 * Returns true if there is a tail node, false otherwise.
	 *
	 * @returns {boolean}
	 * @memberof DoublyLinkedList
	 */
	public hasTail(): boolean {
		return this._tail != null
	}

	/**
	 * Returns true if there is a head node, false otherwise.
	 *
	 * @returns {boolean}
	 * @memberof DoublyLinkedList
	 */
	public hasHead(): boolean {
		return this._head != null
	}
}
