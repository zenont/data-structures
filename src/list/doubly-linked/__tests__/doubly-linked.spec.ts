import { DoublyLinkedList } from '../doubly-linked'

describe('doubly linked List', () => {
	it('should create', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)

		expect(linkedList).toMatchSnapshot()
	})

	it('should create from array', () => {
		const linkedList = DoublyLinkedList.of(1, 2, 3, 4, 5)
		expect(linkedList).toMatchSnapshot()
	})

	it('should create empty', () => {
		const linkedList = DoublyLinkedList.of()
		expect(linkedList).toMatchSnapshot()
	})

	it('should push one', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		const created = linkedList.push(2)

		expect(linkedList).toMatchSnapshot()
	})

	it('should push many', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)

		expect(linkedList).toMatchSnapshot()
	})

	it('should remove middle', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		const toRemove = linkedList.tail()
		linkedList.push(4)
		linkedList.push(5)

		linkedList.remove(toRemove)
		expect(linkedList).toMatchSnapshot()
	})

	it('should remove head', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		const toRemove = linkedList.head()

		linkedList.remove(toRemove!)
		expect(linkedList).toMatchSnapshot()
	})

	it('should remove tail', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		const toRemove = linkedList.tail()

		linkedList.remove(toRemove!)
		expect(linkedList).toMatchSnapshot()
	})

	it('should remove all items', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		let toRemove = linkedList.tail()
		linkedList.remove(toRemove!)
		toRemove = linkedList.tail()
		linkedList.remove(toRemove!)
		toRemove = linkedList.tail()
		linkedList.remove(toRemove!)
		expect(linkedList).toMatchSnapshot()
	})

	it('should pop item', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		const popped = linkedList.pop()
		expect(popped).toMatchSnapshot()
		expect(linkedList).toMatchSnapshot()
	})

	it('should pop all items', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		let popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		expect(popped).toMatchSnapshot()
		expect(linkedList).toMatchSnapshot()
	})

	it('should over pop', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.push(2)
		linkedList.push(3)
		linkedList.push(4)
		linkedList.push(5)
		let popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		popped = linkedList.pop()
		expect(popped).toMatchSnapshot()
		expect(linkedList).toMatchSnapshot()
	})

	it('should return true for head', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		expect(linkedList.hasHead()).toEqual(true)
	})

	it('should return false for head', () => {
		const seed = 1
		const linkedList = DoublyLinkedList.of(seed)
		linkedList.pop()
		expect(linkedList.hasHead()).toEqual(false)
	})

	it('should push after', () => {
		const linkedList = DoublyLinkedList.of(1)
		linkedList.pushAfter(linkedList.tail(), 2)
		expect(linkedList).toMatchSnapshot()
	})

	it('should push after node in the middle', () => {
		const linkedList = DoublyLinkedList.of(1, 2)
		const after = linkedList.tail()
		linkedList.push(4)
		linkedList.pushAfter(after, 3)
		expect(linkedList).toMatchSnapshot()
	})
})
