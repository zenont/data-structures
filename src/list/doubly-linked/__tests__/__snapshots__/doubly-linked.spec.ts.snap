// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doubly linked List should create 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": undefined,
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": undefined,
    "value": 1,
  },
}
`;

exports[`doubly linked List should create empty 1`] = `
DoublyLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`doubly linked List should create from array 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": DoublyLinkedListItem {
        "next": DoublyLinkedListItem {
          "next": DoublyLinkedListItem {
            "next": undefined,
            "prev": [Circular],
            "value": 5,
          },
          "prev": [Circular],
          "value": 4,
        },
        "prev": [Circular],
        "value": 3,
      },
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": DoublyLinkedListItem {
        "next": [Circular],
        "prev": DoublyLinkedListItem {
          "next": [Circular],
          "prev": DoublyLinkedListItem {
            "next": [Circular],
            "prev": undefined,
            "value": 1,
          },
          "value": 2,
        },
        "value": 3,
      },
      "value": 4,
    },
    "value": 5,
  },
}
`;

exports[`doubly linked List should over pop 1`] = `undefined`;

exports[`doubly linked List should over pop 2`] = `
DoublyLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`doubly linked List should pop all items 1`] = `1`;

exports[`doubly linked List should pop all items 2`] = `
DoublyLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`doubly linked List should pop item 1`] = `5`;

exports[`doubly linked List should pop item 2`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": DoublyLinkedListItem {
        "next": DoublyLinkedListItem {
          "next": undefined,
          "prev": [Circular],
          "value": 4,
        },
        "prev": [Circular],
        "value": 3,
      },
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": DoublyLinkedListItem {
        "next": [Circular],
        "prev": DoublyLinkedListItem {
          "next": [Circular],
          "prev": undefined,
          "value": 1,
        },
        "value": 2,
      },
      "value": 3,
    },
    "value": 4,
  },
}
`;

exports[`doubly linked List should push after 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": undefined,
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": undefined,
      "value": 1,
    },
    "value": 2,
  },
}
`;

exports[`doubly linked List should push after node in the middle 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": DoublyLinkedListItem {
        "next": DoublyLinkedListItem {
          "next": undefined,
          "prev": [Circular],
          "value": 4,
        },
        "prev": [Circular],
        "value": 3,
      },
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": DoublyLinkedListItem {
        "next": [Circular],
        "prev": DoublyLinkedListItem {
          "next": [Circular],
          "prev": undefined,
          "value": 1,
        },
        "value": 2,
      },
      "value": 3,
    },
    "value": 4,
  },
}
`;

exports[`doubly linked List should push many 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": DoublyLinkedListItem {
        "next": DoublyLinkedListItem {
          "next": DoublyLinkedListItem {
            "next": undefined,
            "prev": [Circular],
            "value": 5,
          },
          "prev": [Circular],
          "value": 4,
        },
        "prev": [Circular],
        "value": 3,
      },
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": DoublyLinkedListItem {
        "next": [Circular],
        "prev": DoublyLinkedListItem {
          "next": [Circular],
          "prev": DoublyLinkedListItem {
            "next": [Circular],
            "prev": undefined,
            "value": 1,
          },
          "value": 2,
        },
        "value": 3,
      },
      "value": 4,
    },
    "value": 5,
  },
}
`;

exports[`doubly linked List should push one 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": undefined,
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": undefined,
      "value": 1,
    },
    "value": 2,
  },
}
`;

exports[`doubly linked List should remove all items 1`] = `
DoublyLinkedList {
  "_head": undefined,
  "_tail": undefined,
}
`;

exports[`doubly linked List should remove head 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": DoublyLinkedListItem {
        "next": DoublyLinkedListItem {
          "next": undefined,
          "prev": [Circular],
          "value": 5,
        },
        "prev": [Circular],
        "value": 4,
      },
      "prev": [Circular],
      "value": 3,
    },
    "prev": undefined,
    "value": 2,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": DoublyLinkedListItem {
        "next": [Circular],
        "prev": DoublyLinkedListItem {
          "next": [Circular],
          "prev": undefined,
          "value": 2,
        },
        "value": 3,
      },
      "value": 4,
    },
    "value": 5,
  },
}
`;

exports[`doubly linked List should remove middle 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": DoublyLinkedListItem {
        "next": DoublyLinkedListItem {
          "next": undefined,
          "prev": [Circular],
          "value": 5,
        },
        "prev": [Circular],
        "value": 4,
      },
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": DoublyLinkedListItem {
        "next": [Circular],
        "prev": DoublyLinkedListItem {
          "next": [Circular],
          "prev": undefined,
          "value": 1,
        },
        "value": 2,
      },
      "value": 4,
    },
    "value": 5,
  },
}
`;

exports[`doubly linked List should remove tail 1`] = `
DoublyLinkedList {
  "_head": DoublyLinkedListItem {
    "next": DoublyLinkedListItem {
      "next": DoublyLinkedListItem {
        "next": DoublyLinkedListItem {
          "next": undefined,
          "prev": [Circular],
          "value": 4,
        },
        "prev": [Circular],
        "value": 3,
      },
      "prev": [Circular],
      "value": 2,
    },
    "prev": undefined,
    "value": 1,
  },
  "_tail": DoublyLinkedListItem {
    "next": undefined,
    "prev": DoublyLinkedListItem {
      "next": [Circular],
      "prev": DoublyLinkedListItem {
        "next": [Circular],
        "prev": DoublyLinkedListItem {
          "next": [Circular],
          "prev": undefined,
          "value": 1,
        },
        "value": 2,
      },
      "value": 3,
    },
    "value": 4,
  },
}
`;
